### ISSUE TITLE
operations on integer SIMD vectors

### ISSUE THREAD

**Issue by 9il** [ldc-developers/ldc#1482]
LDC allows to multiply  vectors composed of `ushort`, but `uint` and `ulong` are not allowed. AVX-512 has `vpmullq` and `vpaddq` instructions, which are supported in LLVM.

---

``` d
void ok(ref __vector(ushort[16]) a)
{
    a = a * a;
}
```

``` asm
    .cfi_startproc
    vmovdqu16   (%rdi), %ymm0
    vpmullw %ymm0, %ymm0, %ymm0
    vmovdqu16   %ymm0, (%rdi)
    retq
```

---

``` d
void fail(ref __vector(ulong[4]) a)
{
    a = a * a;
}
```

```
  avx                   - Enable AVX instructions.
  avx2                  - Enable AVX2 instructions.
  avx512bw              - Enable AVX-512 Byte and Word Instructions.
  avx512cd              - Enable AVX-512 Conflict Detection Instructions.
  avx512dq              - Enable AVX-512 Doubleword and Quadword Instructions.
  avx512er              - Enable AVX-512 Exponential and Reciprocal Instructions.
  avx512f               - Enable AVX-512 instructions.
  avx512pf              - Enable AVX-512 PreFetch Instructions.
  avx512vl              - Enable AVX-512 Vector Length eXtensions.
```

**Comment by JohanEngelen** [comment]
may sound stupid but: small code example please! ;)

**Comment by 9il** [comment]
done)

**Comment by JohanEngelen** [comment]
There is a hardcoded error upon anything other than multiplying vectors of ushort or short... Is there any reason why you'd want an error on code like this?
```D
void ok(ref __vector(uint[64]) a)
{
    a = a * a;
}
```
Sure, it won't be a nice single instruction multiply but it will work fine. 
Working on a PR.

**Comment by JohanEngelen** [comment]
There are some more hardcoded restrictions on operations with __vector, so you may run into more troubles. Just open a new issue for them ;)

**Comment by redstar** [comment]
Please be aware: the restrictions are by design! We should not change the language in an uncontrolled way. See my comment in the PR. We need a generic solution here, not just LDC hacks.

**Comment by JohanEngelen** [comment]
What was the design rationale? (is this one other case of "DMD is the spec" ?)
I would expect the code to run on any processor, regardless of whether it has special instructions for it. GCC spec was what I was expecting:
"Specifying a combination that is not valid for the current architecture causes GCC to synthesize the instructions using a narrower mode. For example, if you specify a variable of type V4SI and your architecture does not allow for this specific SIMD type, GCC produces code that uses 4 SIs."
https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html
That is also what LLVM appears to do.

**Comment by 9il** [comment]
This is not language change. IMHO

**Comment by redstar** [comment]
The design rationale of DMD is: only those operations are allowed which can be mapped to native SSE instructions. Synthesizing is disallowed. The reason is that you should know upfront what you will get. With LDC, you have to use a disassembler to check.

IMHO the LDC/GCC approach is the right one because you still have to measure to get the performance you want.

I am a bit reluctant to make arbitrary changes to the frontend. @JohanEngelen is right with his comment that there are a lot of other hardcoded restrictions, too. We should make a list of required changes and then create a PR that is somewhat customizable. Switched on for LDC/GCC, switched off for DMD.

**Comment by markos** [comment]
As this is something that affects me a bit (SIMD programming in general), I can't help but commenting here. Depending too much on SSE support -and only that if I understand what you mean- is not very good on portability. I'm currently in the process of writing a SIMD book/library and I am in a wait-state with regards to D/LDC, I'd love to write about it, but it's too much x86-centric right now. A language should not make any decisions on type and operations support taking into account only one architecture -even if it's the most popular one. Right now, I'm writing SIMD code for x86, ARM (32 and 64-bit), PowerPC (Both Altivec/VSX and both big/little endian), and also very recently added s390x ZVector support -and in the not too distant future I plan to add MIPS SIMD support as well. I've postponed doing it in D as right now core.simd in just SSE. I would love doing it in LDC low level intrinsics, but it's too much compiler-centric. I'd love to be able to use the same C intrinsics in D as well and be done with it, I really dislike the opcodes->method system that's used in core.simd. Comments welcome.

**Comment by redstar** [comment]
Stupid idea: Why not create a flag `Globals.relaxVectorTypes` and use this in the frontend code? With default `false` it is the DMD behaviour, with `true` we change the type changes to our relaxed rules. (`relax` may not be the right word for it)
I then would try to get this merge upstream.

**Comment by JohanEngelen** [comment]
Btw, dlang.org provides a workaround for if the compiler does not support operations:
```D
float4 a,b;
static if (__traits(compiles, a/b))
    c = a / b;
else
    c[] = a[] / b[];
```
DMD could do the rewriting itself, rather than having the programmer do it.

**Comment by 9il** [comment]
> DMD could do the rewriting itself, rather than having the programmer do it.

Plus it should completely unroll the loop

**Comment by JohanEngelen** [comment]
> The reason is that you should know upfront what you will get. 

I don't think this makes sense. Writing `long b = a * 2` you also don't know whether you will get a mul, add, shift, whatever instructions and why would you care? In the current state, to me it seems __vector is completely useless: compilation can break easily for any target, and so you have to write the alternative with `[]` everywhere. 

> Stupid idea: Why not create a flag Globals.relaxVectorTypes and use this in the frontend code? 

Would work for LDC.

**Comment by 9il** [comment]
> The reason is that you should know upfront what you will get.
> I don't think this makes sense. Writing long b = a * 2 you also don't know whether you will get a mul, add, shift, whatever instructions and why would you care? In the current state, to me it seems __vector is completely useless: compilation can break easily for any target, and so you have to write the alternative with [] everywhere.

In addition current LDC allows to use vectors of larger size: they are represented using multiple avx registers

**Comment by redstar** [comment]
@markos Which intrinsics do you use? The gcc builtins are available in LDC, too. (And I am curious about your book. Great project!)

The opcodes->method system is only supported by DMD. I also do not like it....

**Comment by markos** [comment]
@redstar which intrinsics? all of them :)
I've done the ARM/PowerPC/S390X ports of Eigen library (using C intrinsics but also inline asm in some cases (eg. compiler bug) and am also in the (very very slow) process of writing a SIMD book that tries to provide a comparative guide between all major SIMD engines. The goal is to produce a language agnostic huge dataset and from that automatically produce any language bindings (but also documentation) that use those intrinsics (to be honest, I tried to crowdfund it at some point but I'm not really good at marketing myself :). Anyway, this is a slow but ongoing process and one I plan on releasing under an open source license soon.

**Comment by redstar** [comment]
@markos LDC supports the gcc builtins in `ldc.gccbuiltins_<arch>`. Therefore I asked.

Your book project sounds really interesting. But a lot of work...

**Comment by 9il** [comment]
@markos are you interesting in BLAS in D? I am working on kernels now, [see also code example](https://github.com/libmir/mir/pull/210)

**Comment by markos** [comment]
@9il I was actually thinking of doing a port of Eigen in D. But, the SIMD situation led me to postpone this. If I could use the gcc builtins then that would work, but it would still mean it will only work only on LDC (or maybe GDC as well?), but most likely not DMD. So that will not be a universal library.

**Comment by 9il** [comment]
> @9il I was actually thinking of doing a port of Eigen in D. But, the SIMD situation led me to postpone this. If I could use the gcc builtins then that would work, but it would still mean it will only work only on LDC (or maybe GDC as well?), but most likely not DMD. So that will not be a universal library.

The proposed kernel is generic and will work with DMD (but without SIMD optimization). However this is not big problem, because main optimization is cache blocking. Btw, Mir compiled with LDC would be faster then Eigen. Current generic kernels looks like [this](https://gist.github.com/9il/58c1b80110de2db5f2eff6999346a928), after #1472 it will have FMA instructions. The good point about Mir's BLAS, that is will have optimal SIMD (with LDC) and cache optimization for any architecture (including arm of course)

**Comment by redstar** [comment]
@JohanEngelen I think I found the right solution. The language specification says about vectors: "These types and operations will be the ones defined for the architecture the compiler is targeting." This is not implemented, instead the x86 solution is hard coded.
What needs to be done is extending `target.d` with `isVectorTypeLegal()` and `isVectorOperationLegal()`. The semantic analysis must then ask these new predicates and act accordingly. The current DMD rules move into these new predicates. We have our own implementation of `target.d`. We would then simply allow all operations because we can't do anything else.
I think this would be a clean solution which can be upstreamed, too.

**Comment by yebblies** [comment]
That sounds fine to me...  It would be nice if all three compilers were doing the same thing though.

**Comment by klickverbot** [comment]
> It would be nice if all three compilers were doing the same thing though.

That's impossible with DMD's lack of related codegen.

**Comment by yebblies** [comment]
Codegen can be improved.  I guess what I mean is that it would be better if Walter could be convinced that all vector types and operations should be allowed.

**Comment by redstar** [comment]
The argument from Walter was: If you allow vector operation on all types then you have to synthesize the ones not natively supported. This is kind of performance trap. To prevent the developer to fall into this trap, only supported operations are allowed. This was considered to be important for game development.

**Comment by yebblies** [comment]
I know... what's the counter-argument?  Can we add `-vsimd` to yell about it and let the default slowly synthesize?

Or is this not something anyone will ever want with dmd?

**Comment by JohanEngelen** [comment]
> This is kind of performance trap.

Why would the performance of "synthesized" ops be worse than whatever workaround code the dev would write?

**Comment by redstar** [comment]
Complexity for the developer. The language changes depending on the CPU model you are compiling for. SSE, AVX, AVX2 - they allow all different types and operators. IMHO the better way to support this is function multiversioning.

**Comment by 9il** [comment]
Why we need need multiversioning ? Just traits `compiles` seems to be enough

**Comment by redstar** [comment]
The difference is that the developer knows about it if he has to write the workaround.

**Comment by JohanEngelen** [comment]
With the same argumentation, you would have to add a check for "multiply" in the D language, and emit an error if the processor cannot do multiplication, or ulong+ulong on a 32bit machine, etc.

**Comment by redstar** [comment]
> Why we need need multiversioning ? Just traits compiles seems to be enough

Traits compiles only works in this case if there are types/operators which are sometimes allowed and somtimes disallowed. In case of LDC it will always compile because LLVM synthesizes the operation.
As an idea multiversioning may rise the awareness of the developer.

**Comment by redstar** [comment]
It's not my argument... :-)

**Comment by JohanEngelen** [comment]
To check what the target can do, __traits(targetHasFeature,...) is much better than writing some code and watching it error, no? :)

**Comment by kinke** [comment]
This works now with LDC 1.5: https://godbolt.org/g/qNy7pm

The ulong version doesn't use `vpmullq` though.

**Comment by 9il** [comment]
Thank you!
